### ISSUE TITLE
NameError: name '__beartype_random_int' is not defined

### ISSUE THREAD

**Issue by albanie** [beartype/beartype#5]
Since the glorious update to 0.3, I encountered a slight issue with using lists of pathlib Path objects.  I'm not sure whether it is specific to pathlib Paths in particular, but this was what I whittled it down to as a minimal script that produces the issue:

```
from typing import List
from pathlib import Path

from beartype import beartype


class UserDefinedType:

    def __repr__(self):
        return "I am a user-defined type"


@beartype
def working_example(x: List[Path]):
    print(f"working_example >>> I will get to see {x}")


@beartype
def working_combination(x: List[int], y: int):
    print(f"working_combination >>> I get to see {x} and {y}")


@beartype
def working_combination_udt(x: List[UserDefinedType], y: int):
    print(f"working_combination with UDT >>> I get to see {x} and {y}")


@beartype
def not_working_combination(x: List[Path], y: int):
    print(f"not_working_combination: Sadly, I will never see {x} or {y}")


if __name__ == "__main__":
    path_list = [Path("/a/b")]
    int_list = [0]
    udt_list = [UserDefinedType()]

    # works OK
    working_example(x=path_list)
    working_combination(x=int_list, y=0)
    working_combination_udt(x=udt_list, y=0)

    # breaks
    not_working_combination(x=path_list, y=0)
```

This produces the output:

```
working_example >>> I will get to see [PosixPath('/a/b')]
working_combination >>> I get to see [0] and 0
working_combination with UDT >>> I get to see [I am a user-defined type] and 0
Traceback (most recent call last):
  File "debug.py", line 43, in <module>
    not_working_combination(x=path_list, y=0)
  File "<string>", line 25, in __beartyped_not_working_combination
NameError: name '__beartype_random_int' is not defined
```

It's easy for me to work around for now by reverting to using `list` as an annotation (rather than `List[Path]`), but just thought I'd flag it as a potential issue.

**Comment by leycec** [comment]
**Gulp.** Looks like glorious 0.3.0 was a scam, folks. In this shocking new issue, citizen code journalist @albanie reveals the dark festering truth at the heart of long-denied `beartype` allegations of fragile, error-prone, and frankly broken type-checking.

Just "Ugh!" Our unit tests were exhaustive, or so the documentation claimed. Although `pathlib.Path` *is* just a non-`typing` class, it's a bit non-standard in that it overrides the `__new__()` dunder method to conditionally return an instance of a different class depending on platform.

I am rolling up my white-collar sleeves and shaking my head ruefully as we speak.

**Comment by leycec** [comment]
**Stranger Things: Python Edition.** I just added a unit test exhaustively exercising `List[Path]`, but... *it passes on all major Python versions.* Something horrifying is happening elsewhere and I don't know what it is.

Let's try running your exact MLE and see what jump scare pops up! :fearful:

**Comment by leycec** [comment]
*"Uh oh."* Horrifyingly, this appears to have nothing to do with `pathlib.Path` types. Instead, there appears to be a critical order-of-function-declarations issue in `beartype` 0.3.0 that I can only replicate with your exact MLE – *and the types don't matter.* I replaced `List[Path]` with `List[int]` in `not_working_combination()` and received the same exception.

My momentary advice is to temporarily revert to `beartype` 0.2.0 until I get the fast CVE hotfix out. I'm sweating armpit-staining bullets. Emoji fear intensifies. :cold_sweat:

**Comment by leycec** [comment]
**Ah-ha!** It's an obscure function memoization issue. You *really* don't want to know. This is why friends don't let friends memoize drunk, people.

I'll have this patched up and `beartype 0.4.1` pushed out the cabin door in a jiffy.

**Comment by leycec** [comment]
**Ka-blammo!** `beartype 0.3.1` is up on all the usual suspects. This turned out to be a surprisingly involved issue. On the bright side, we've resolved a critical memoization defect and written unit tests exhaustively exercising this defect across all possible edge cases. On the dark side, this should never have happened.

But that's 2020 for you. *Urgh!*
